java/arcs/core/tools/Proto2Schema.kt

```diff
@@ -0,0 +1,138 @@
+                        }
+                    }
+                    .unindent()
+                    .addStatement("),")
+                    .addStatement("%T(", schemaFieldsClass)
+                    .indent()
+                    .addStatement("singletons = mapOf(")
+                    .indent()
+                    .apply {
+                        val entries = it.fields.singletons.entries
+                        entries.forEachIndexed { index, entry ->
+                            when (entry.value.tag) {
+                                FieldType.Tag.EntityRef -> add(
+                                    "%S to %T(%S)",
+                                    entry.key,
+                                    FieldType.EntityRef::class,
+                                    (entry.value as FieldType.EntityRef).schemaHash
+                                )
+                                FieldType.Tag.Primitive -> add(
+                                    "%S to %T.%L",
+                                    entry.key,
+                                    FieldType::class,
+                                    (entry.value as FieldType.Primitive).primitiveType
+                                )
+                            }
+                            if (index != entries.size - 1) add(",")
+                            add("\n")
+                        }
+                    }
+                    .unindent()
+                    .addStatement("),")
+                    .addStatement("collections = mapOf(")
+                    .indent()
+                    .apply {
+                        val entries = it.fields.collections.entries
+                        entries.forEachIndexed { index, entry ->
+                            when (entry.value.tag) {
+                                FieldType.Tag.EntityRef -> add(
+                                    "%S to %T(%S)",
+                                    entry.key,
+                                    FieldType.EntityRef::class,
+                                    (entry.value as FieldType.EntityRef).schemaHash
+                                )
+                                FieldType.Tag.Primitive -> add(
+                                    "%S to %T.%L",
+                                    entry.key,
+                                    FieldType::class,
+                                    (entry.value as FieldType.Primitive).primitiveType
+                                )
+                            }
+                            if (index != entries.size - 1) add(",")
+                            add("\n")
+                        }
+                    }
+                    .unindent()
+                    .addStatement(")")
+                    .unindent()
+                    .addStatement("),")
+                    .addStatement("%S", it.hash)
+                    .addStatement(")")
+                    .build())
```

Comment:
This is way too much logic to exist in one monolith. Please break this up into smaller, more-teestable and more-readable functions.